### ISSUE TITLE
time stamp for reconstructed data

### ISSUE THREAD

**Issue by hofmannmartin** [MagneticParticleImaging/MDF#72]
I think it would be a good idea to have the posibility to provide time stamps for the reconstructed data to assign frames to specific time points. This information is required if one want to create a movie from the reconstructed data or if one wants to correlate the MPI data with other parallel obtained measurements.

**Comment by NeumannIMT** [comment]
Aren't dimensions N (frames) and J (patches) strictly consecutive in time (including a fixed delay)? 
If not a delay/timestamp vector/matrix is indeed needed to describe the temporal connection between those. 
The next question then is: How should the timestamp be stored and how precise must it be (ms, µs, ns)? (The way the time is saved for the file feels inappropiate for it)

Are overlapping (overlapping in time) Frames/Patches allowed? (If even possible?)

**Comment by hofmannmartin** [comment]
If the frames N are consecutive or not depends on the flags `isFramePermutation`.

But that is for the raw data. The reconstructed sequence of MPI tomgrams is not necessarily in a one to one correspondence to the raw frames. It could be that the data was averaged or otherwise post processed before reconstruction. That is why we have O instead of N reconstructed frames in general. Hence one can not infer time differences or absolute time for the reconstruced frames.

To describe the time dimension one could have one time stamp for the first reconstructed frame and time differences between consecutive ones for example.

**Comment by tknopp** [comment]
We store this in the recon params, which we at IBI store as extension to the MDF. This should be the starting point for discussions how to extend the MDF in this area. We currently store the reconstructed frames and the averaging factor.

I am fine with standardizing these parameters. There is also an issue about that.

**Comment by tknopp** [comment]
https://github.com/MagneticParticleImaging/MDF/issues/46

**Comment by tknopp** [comment]
Frames is tricky by the way. We store ranges here (our UI uses ranges). Requires hdf5 attributes

**Comment by hofmannmartin** [comment]
I would like to have more flexibility. If for example we would like to pipe the online reconstruction results to an MDF File, we will have an issue with varying delays between frames.

**Comment by NeumannIMT** [comment]
> If the frames N are consecutive or not depends on the flags isFramePermutation. But that is for the raw data.

So the raw data also needs timestamp? Maybe we should open an issue for that? And maybe from there we can than derive a timestamp used for a reconstructed frame?

> Frames is tricky by the way. We store ranges here (our UI uses ranges). Requires hdf5 attributes

Although I really like the idea of using attributes since all the is<Something> Flags could be attributes to a dataset; please dont use them. (No real reason, just more coding to get the abstract maping to an attribute correct. Maping to a dataset is basically free since the code already exists)

Concerning: #45 #46
I would really like to see that mdf defines subfiles for different things more strictly (measurement, reconstruction, systemmatrix/calibration) and then defines an vmdf (virtual mdf) to virtually link all those together. Instead of one file format which tries to do it all with a lot of optional fields.

**Comment by hofmannmartin** [comment]
> So the raw data also needs timestamp? Maybe we should open an issue for that? And maybe from there we can than derive a timestamp used for a reconstructed frame?

Since the `/measurement/framePermutation` and the `/acquisition/startTime` is known the times for each frame can be derived.

>Concerning: #45 #46
I would really like to see that mdf defines subfiles for different things more strictly (measurement, reconstruction, systemmatrix/calibration) and then defines an vmdf (virtual mdf) to virtually link all those together. Instead of one file format which tries to do it all with a lot of optional fields.

That sounds like an interesting idea. Can you provide a more detailed explanation of what the differences are?

**Comment by tknopp** [comment]
Raw data is continuous. No timestamps needed.

Martin: recon frames are not necessarily ranges: we also store index vectors. Think about Flos robot measurements.

Alexander seems to be concerned about redundantly storing metadata in measurement and in reco files. But this is absolutely necessary. An MDF should be self contained.

Regarding #45 (and also #46). We use this system very successful "in production". The code for all this is in MPIFiles.jl. Our database is already quite large, so we will certainly not change this, without a serious need.

**Comment by NeumannIMT** [comment]
> Raw data is continuous. No timestamps needed.

Are you absolutely sure about this? Especially patches could have hardware specific problems for continuous acquisition. Meaning depending on how far away patches are the hardware needs different delays to really reach the patch region. (But I could be wrong in my understanding of patches which is based on what Mandy did in her thesis.)

Maybe a better definition of frame and patches is needed in the mdf to answer my questions:
- When is a frame considered a new measurement?
- When is a new measurement of the same imaging plane and dimensions not considered a frame?
- When is a patch considered a frame? (Is a patch not always a smaller frame?)
- Are patches allowed to be taken in arbitary order and time?
- Was there already a discussion about that? If yes please link it.

Or is there a hard requirement for the N x J dimensions to be strictly continous in time meaning continous acquisition without any pause? If yes, it is basically a defining characteristic for frames and patches as it is understood by an mdf and must be included in the mdf specification.  

Considering patches the mdf currently states:
> In a multi-patch setting J offsetFields or mechanical movements shift the gradient field by offsetFieldShift, where J is the number of patches numPatches of a multi-patch measurement. For
each shift at least one full drive field cycle is used to acquire the measurement data.

From that definition it looks like it does not have to be continous in time. 

Considering the definition of frames:
> Along this dimension, the frames are ordered with respect to the time at which they were acquired starting with the measurement acquired first and stopping with the measurement acquired last

From that definition I am also allowed to have variable time delays between frames. 

-> should I open a new Issue: Timestamp for frames and patches? Seems like that this must be discussed first before we can answer the timestamps for the reconstruction answer but it seems like mdf will need them.

> That sounds like an interesting idea. Can you provide a more detailed explanation of what the differences are?

> Alexander seems to be concerned about redundantly storing metadata in measurement and in reco files. But this is absolutely necessary. An MDF should be self contained.
Regarding #45 (and also #46). We use this system very successful "in production". The code for all this is in MPIFiles.jl. Our database is already quite large, so we will certainly not change this, without a serious need.

Will explain it later in a different issue with a bit more detail (Is there another way to open a discussion instead of opening a issues?). I am not really concerned about metadata because metadata is generally small compared to everything else in MPI. If you want to have a bit more insight now look at [HDF5 Virtual Datasets](https://support.hdfgroup.org/HDF5/docNewFeatures/NewFeaturesVirtualDatasetDocs.html). The PDF at the end of the page gives a good overview how it might be used. (Also the mdf format already recommends to use 3 different subfiles for each type, i would just go further make it required instead of a recommendation an define 3 formats. The next step then ist to describe the dependencies of the files and virtually link them together)

**Comment by tknopp** [comment]
Discussions are initiated by opening an issue. That's the regular github workflow

**Comment by tknopp** [comment]
The data during field switching is also stored. We should improve documentation to make this clear. It may make sense to drop the data. Not sure. Since we have no scanner supporting FF this discussion is pretty theoretical.
